### PR TITLE
Impro 1569 apply emos land sea

### PR DESCRIPTION
**Pull request by fionaRust**
Addresses the last points of IMPRO-1569:
* Ensures that EnsembleReordering plugin handles masked data.
* Adds a function for combining calibrated land data with uncalibrated sea data.
* Adds CLI tests using the land sea option for apply-emos-coefficients CLI - a notebook showing the new data will follow

There were a couple of other minor changes I made, which are in separate commits so it's easier to see what they are.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/cli/apply_emos_coefficients.py  1
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  2
         

Clones added
============
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  4
- improver_tests/acceptance/test_apply_emos_coefficients.py  2
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  1
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5059910&bid=16653466)

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py [hunk 1]
```diff
@@ -553,5 +556,85 @@ def test_missing_coordinate(self):
...
+            [[[3.0, 1.0, 1.0, 1.0],
+              [3.0, 1.0, 1.0, 1.0],
+              [1.0, 1.0, 1.0, 3.0]],
+             [[3.0, 1.0, 1.0, 1.0],
+              [3.0, 1.0, 1.0, 1.0],
+              [1.0, 1.0, 1.0, 3.0]]], dtype=np.float32)
+        expected_cube = self.percentiles_land.copy()
+        expected_cube.data = expected_merged_data
+        merge_land_and_sea(self.percentiles_land, self.percentiles_sea)
+        self.assertArrayAlmostEqual(
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py]: I would have thought that ```numpy.testing.assert_array_equal``` would work here. There is nothing approximate about this, we haven't modified the values at all.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py [hunk 2]
```diff
@@ -553,5 +556,85 @@ def test_missing_coordinate(self):
...
+        self.assertFalse(np.ma.is_masked(self.percentiles_land.data))
+        self.assertEqual(self.percentiles_land.data.dtype, np.float32)
+
+    def test_nothing_to_merge(self):
+        """Test case where there is no missing data to fill in."""
+        input_mask = np.ones((2, 3, 4)) * False
+        self.percentiles_land.data.mask = input_mask
+        expected_cube = self.percentiles_land.copy()
+        merge_land_and_sea(self.percentiles_land, self.percentiles_sea)
+        self.assertArrayAlmostEqual(
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py]: Equivalence could be used here as well.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py [hunk 3]
```diff
@@ -553,5 +556,85 @@ def test_missing_coordinate(self):
...
+            self.percentiles_land.xml(checksum=True))
+        self.assertFalse(np.ma.is_masked(self.percentiles_land.data))
+        self.assertEqual(self.percentiles_land.data.dtype, np.float32)
+
+    def test_input_not_masked(self):
+        """Test case where input cube is not masked."""
+        self.percentiles_land.data = np.ones((2, 3, 4), dtype=np.float32)
+        expected_cube = self.percentiles_land.copy()
+        merge_land_and_sea(self.percentiles_land, self.percentiles_sea)
+        self.assertArrayAlmostEqual(
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py]: And again.

File: improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py [hunk 4]
```diff
@@ -420,6 +420,85 @@ def test_3d_cube(self):
...
+            [[[1, 1]],
+             [[2, 2]],
+             [[3, 3]]], mask=mask, dtype=np.float32)
+
+        # Reordering of the calibrated_data array to match
+        # the raw_data ordering
+        result_data = np.array(
+            [[[np.nan, 1]],
+             [[np.nan, 2]],
+             [[np.nan, 3]]], dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py]: I did think about this at the time and was going to swap it so the other column was masked. I think your data is better though so I've changed it as you suggested.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  2
- improver/cli/apply_emos_coefficients.py  1
         

Clones added
============
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  1
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  16
- improver_tests/acceptance/test_apply_emos_coefficients.py  2
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_utilities.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5059910&bid=16653466)

File: improver/ensemble_copula_coupling/utilities.py [hunk 5]
```diff
@@ -313,3 +313,34 @@ def restore_non_probabilistic_dimensions(
     shape_to_reshape_to = (
         [output_probabilistic_dimension_length] + shape_to_reshape_to)
     return array_to_reshape.reshape(shape_to_reshape_to)
+
+
+def merge_land_and_sea(calibrated_land_only, uncalibrated):
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver/ensemble_copula_coupling/utilities.py]: Done

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  2
- improver/cli/apply_emos_coefficients.py  1
         

Clones added
============
- improver_tests/acceptance/test_apply_emos_coefficients.py  2
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_RebadgePercentilesAsRealizations.py  1
- improver_tests/calibration/utilities/test_utilities.py  4
- improver_tests/ensemble_copula_coupling/ensemble_copula_coupling/test_EnsembleReordering.py  16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5059910&bid=16653466)

**Pull request merged**